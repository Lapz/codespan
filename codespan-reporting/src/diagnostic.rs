//! Diagnostic data structures.

use codespan::{FileId, Span};
#[cfg(feature = "serialization")]
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;

/// A severity level for diagnostic messages.
///
/// These are ordered in the following way:
///
/// ```rust
/// use codespan_reporting::diagnostic::Severity;
///
/// assert!(Severity::Bug > Severity::Error);
/// assert!(Severity::Error > Severity::Warning);
/// assert!(Severity::Warning > Severity::Note);
/// assert!(Severity::Note > Severity::Help);
/// ```
#[derive(Copy, Clone, PartialEq, Hash, Debug)]
#[cfg_attr(feature = "serialization", derive(Serialize, Deserialize))]
pub enum Severity {
    /// An unexpected bug.
    Bug,
    /// An error.
    Error,
    /// A warning.
    Warning,
    /// A note.
    Note,
    /// A help message.
    Help,
}

impl Severity {
    /// We want bugs to be the maximum severity, errors next, etc...
    fn to_cmp_int(self) -> u8 {
        match self {
            Severity::Bug => 5,
            Severity::Error => 4,
            Severity::Warning => 3,
            Severity::Note => 2,
            Severity::Help => 1,
        }
    }
}

impl PartialOrd for Severity {
    fn partial_cmp(&self, other: &Severity) -> Option<Ordering> {
        u8::partial_cmp(&self.to_cmp_int(), &other.to_cmp_int())
    }
}

/// A diagnostic title.
#[derive(Clone, Debug)]
#[cfg_attr(feature = "serialization", derive(Serialize, Deserialize))]
pub struct Title {
    /// An optional code that identifies this diagnostic.
    pub code: Option<String>,
    /// The main message associated with this diagnostic.
    ///
    /// These should not include line breaks.
    pub message: String,
}

impl Title {
    /// Create a new diagnostic title.
    pub fn new(message: impl Into<String>) -> Title {
        Title {
            message: message.into(),
            code: None,
        }
    }

    /// Add an error code to the diagnostic title.
    pub fn with_code(mut self, code: impl Into<String>) -> Title {
        self.code = Some(code.into());
        self
    }
}

impl From<String> for Title {
    fn from(message: String) -> Title {
        Title::new(message)
    }
}

impl From<&str> for Title {
    fn from(message: &str) -> Title {
        Title::new(message)
    }
}

#[derive(Clone, Debug)]
#[cfg_attr(feature = "serialization", derive(Serialize, Deserialize))]
pub enum LabelStyle {
    /// Labels that describe the primary cause of a diagnostic.
    Primary,
    /// Labels that provide additional context for a diagnostic.
    Secondary,
}

/// A label describing an underlined region of code associated with a diagnostic.
#[derive(Clone, Debug)]
#[cfg_attr(feature = "serialization", derive(Serialize, Deserialize))]
pub struct Label {
    /// The style of the label.
    pub style: LabelStyle,
    /// The file that we are labelling.
    pub file_id: FileId,
    /// The span we are going to include in the final snippet.
    pub span: Span,
    /// A message to provide some additional information for the underlined
    /// code. These should not include line breaks.
    pub message: String,
}

impl Label {
    /// Create a new label.
    pub fn new(
        style: LabelStyle,
        file_id: FileId,
        span: impl Into<Span>,
        message: impl Into<String>,
    ) -> Label {
        Label {
            style,
            file_id,
            span: span.into(),
            message: message.into(),
        }
    }

    /// Create a new primary label.
    pub fn primary(file_id: FileId, span: impl Into<Span>, message: impl Into<String>) -> Label {
        Label::new(LabelStyle::Primary, file_id, span, message)
    }

    /// Create a new secondary label.
    pub fn secondary(file_id: FileId, span: impl Into<Span>, message: impl Into<String>) -> Label {
        Label::new(LabelStyle::Secondary, file_id, span, message)
    }
}

/// Represents a diagnostic message that can provide information like errors and
/// warnings to the user.
#[derive(Clone, Debug)]
#[cfg_attr(feature = "serialization", derive(Serialize, Deserialize))]
pub struct Diagnostic {
    /// The overall severity of the diagnostic
    pub severity: Severity,
    /// The title of the diagnostic.
    pub title: Option<Title>,
    /// Source labels that describe the cause of the diagnostic.
    pub labels: Vec<Label>,
    /// Notes that are associated with the primary cause of the diagnostic.
    /// These can include line breaks for improved formatting.
    pub notes: Vec<String>,
}

impl Diagnostic {
    /// Create a new diagnostic.
    pub fn new(severity: Severity) -> Diagnostic {
        Diagnostic {
            severity,
            title: None,
            labels: Vec::new(),
            notes: Vec::new(),
        }
    }

    /// Create a new diagnostic with a severity of `Severity::Bug`.
    pub fn bug() -> Diagnostic {
        Diagnostic::new(Severity::Bug)
    }

    /// Create a new diagnostic with a severity of `Severity::Error`.
    pub fn error() -> Diagnostic {
        Diagnostic::new(Severity::Error)
    }

    /// Create a new diagnostic with a severity of `Severity::Warning`.
    pub fn warning() -> Diagnostic {
        Diagnostic::new(Severity::Warning)
    }

    /// Create a new diagnostic with a severity of `Severity::Note`.
    pub fn note() -> Diagnostic {
        Diagnostic::new(Severity::Note)
    }

    /// Create a new diagnostic with a severity of `Severity::Help`.
    pub fn help() -> Diagnostic {
        Diagnostic::new(Severity::Help)
    }

    /// Add a title to the diagnostic.
    pub fn with_title(mut self, title: impl Into<Title>) -> Diagnostic {
        self.title = Some(title.into());
        self
    }

    /// Add some labels to the diagnostic.
    pub fn with_labels(mut self, labels: Vec<Label>) -> Diagnostic {
        self.labels = labels;
        self
    }

    /// Add some notes to the diagnostic.
    pub fn with_notes(mut self, notes: Vec<String>) -> Diagnostic {
        self.notes = notes;
        self
    }
}
