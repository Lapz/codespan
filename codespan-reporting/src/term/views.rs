mod border;
mod gutter;
mod header;
mod locus;
mod new_line;
mod note;
mod source_snippet;
mod underline;

pub use self::border::*;
pub use self::gutter::*;
pub use self::header::*;
pub use self::locus::*;
pub use self::new_line::*;
pub use self::note::*;
pub use self::source_snippet::*;
pub use self::underline::*;
