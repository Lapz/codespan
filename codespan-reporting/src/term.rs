//! Terminal back-end for emitting diagnostics.

use codespan::Files;
use std::io;
use std::str::FromStr;
use termcolor::{ColorChoice, WriteColor};

use crate::diagnostic::{Diagnostic, LabelStyle};

mod config;
mod views;

pub use termcolor;

pub use self::config::{Chars, Config, Styles};

/// Emit a diagnostic using the given writer, context, config, and files.
pub fn emit<Source: AsRef<str>>(
    writer: &mut (impl WriteColor + ?Sized),
    config: &Config,
    files: &Files<Source>,
    diagnostic: &Diagnostic,
) -> io::Result<()> {
    use std::collections::BTreeMap;

    use self::views::MarkStyle;
    use self::views::{Header, NewLine, Note, SourceSnippet};

    // Emit the title
    //
    // ```text
    // error[E0001]: unexpected type in `+` application
    // ```
    if let Some(title) = &diagnostic.title {
        Header::new(diagnostic.severity, title).emit(writer, config)?;
        NewLine::new().emit(writer, config)?;
    }

    // Group labels by file

    let mut label_groups = BTreeMap::new();

    for label in &diagnostic.labels {
        let mark_style = match label.style {
            LabelStyle::Primary => MarkStyle::Primary(diagnostic.severity),
            LabelStyle::Secondary => MarkStyle::Secondary,
        };
        label_groups
            .entry(label.file_id)
            .or_insert(vec![])
            .push((label, mark_style));
    }

    // Emit the snippets, starting with the one that contains the primary label

    for (file_id, labels) in label_groups {
        // FIXME: consistent gutter padding
        SourceSnippet::new(file_id, labels).emit(files, writer, config)?;
    }

    // Additional notes
    //
    // ```text
    // = expected type `Int`
    //      found type `String`
    // ```
    for note in &diagnostic.notes {
        let gutter_padding = 0; // TODO: use gutter padding from emitting source snippets
        Note::new(gutter_padding, &note).emit(writer, config)?;
    }
    NewLine::new().emit(writer, config)?;

    Ok(())
}

/// A command line argument that configures the coloring of the output.
///
/// This can be used with command line argument parsers like `clap` or `structopt`.
///
/// # Example
///
/// ```rust
/// use structopt::StructOpt;
/// use codespan_reporting::term::termcolor::StandardStream;
/// use codespan_reporting::term::ColorArg;
///
/// #[derive(Debug, StructOpt)]
/// #[structopt(name = "groovey-app")]
/// pub struct Opts {
///     /// Configure coloring of output
///     #[structopt(
///         long = "color",
///         parse(try_from_str),
///         default_value = "auto",
///         possible_values = ColorArg::VARIANTS,
///         case_insensitive = true
///     )]
///     pub color: ColorArg,
/// }
///
/// fn main() {
///     let opts = Opts::from_args();
///     let writer = StandardStream::stderr(opts.color.into());
/// }
/// ```
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct ColorArg(pub ColorChoice);

impl ColorArg {
    /// Allowed values the argument.
    ///
    /// This is useful for generating documentation via `clap` or `structopt`'s
    /// `possible_values` configuration.
    pub const VARIANTS: &'static [&'static str] = &["auto", "always", "ansi", "never"];
}

impl FromStr for ColorArg {
    type Err = &'static str;

    fn from_str(src: &str) -> Result<ColorArg, &'static str> {
        match src {
            _ if src.eq_ignore_ascii_case("auto") => Ok(ColorArg(ColorChoice::Auto)),
            _ if src.eq_ignore_ascii_case("always") => Ok(ColorArg(ColorChoice::Always)),
            _ if src.eq_ignore_ascii_case("ansi") => Ok(ColorArg(ColorChoice::AlwaysAnsi)),
            _ if src.eq_ignore_ascii_case("never") => Ok(ColorArg(ColorChoice::Never)),
            _ => Err("valid values: auto, always, ansi, never"),
        }
    }
}

impl Into<ColorChoice> for ColorArg {
    fn into(self) -> ColorChoice {
        self.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::diagnostic::Label;

    #[test]
    fn unsized_emit() {
        let mut files = Files::new();

        let id = files.add("test", "");
        emit(
            &mut termcolor::NoColor::new(Vec::<u8>::new()) as &mut dyn WriteColor,
            &Config::default(),
            &files,
            &Diagnostic::bug().with_labels(vec![Label::primary(id, codespan::Span::default(), "")]),
        )
        .unwrap();
    }
}
