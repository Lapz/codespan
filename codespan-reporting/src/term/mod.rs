//! Terminal back-end for emitting diagnostics.

use codespan::Files;
use std::io;
use std::str::FromStr;
use termcolor::{Color, ColorChoice, ColorSpec, WriteColor};

use crate::diagnostic::{Diagnostic, Severity};

pub use termcolor;

mod views;

/// Emit a diagnostic using the given writer, context, config, and files.
pub fn emit(
    writer: &mut impl WriteColor,
    config: &Config,
    files: &Files,
    diagnostic: &Diagnostic,
) -> io::Result<()> {
    use self::views::{RichDiagnostic, ShortDiagnostic};

    match config.display_style {
        DisplayStyle::Rich => RichDiagnostic::new(files, diagnostic).emit(writer, config),
        DisplayStyle::Short => ShortDiagnostic::new(files, diagnostic).emit(writer, config),
    }
}

/// Configures how a diagnostic is rendered.
#[derive(Clone, Debug)]
pub struct Config {
    /// The display style to use when rendering diagnostics.
    /// Defaults to: `DisplayStyle::Rich`.
    pub display_style: DisplayStyle,
    /// Column width of tabs.
    /// Defaults to: `4`.
    pub tab_width: usize,
    /// Styles to use when rendering the diagnostic.
    pub styles: Styles,
    /// The character to use for the top-left corner of the source.
    /// Defaults to: `'┌'`.
    pub source_border_top_left_char: char,
    /// The character to use for the top border of the source.
    /// Defaults to: `'─'`.
    pub source_border_top_char: char,
    /// The character to use for the left border of the source.
    /// Defaults to: `'│'`.
    pub source_border_left_char: char,
    /// The character to use for marking a primary label.
    /// Defaults to: `'^'`.
    pub primary_caret_char: char,
    /// The character to use for marking a secondary label.
    /// Defaults to: `'-'`.
    pub secondary_caret_char: char,
    /// The character to use for marking the ends of a multi-line primary label.
    /// Defaults to: `'^'`.
    pub multiline_primary_caret_char: char,
    /// The character to use for marking the ends of a multi-line secondary label.
    /// Defaults to: `'\''`.
    pub multiline_secondary_caret_char: char,
    /// The character to use for the top-left corner of a multi-line label.
    /// Defaults to: `'╭'`.
    pub multiline_top_left_char: char,
    /// The character to use for the top of a multi-line label.
    /// Defaults to: `'─'`.
    pub multiline_top_char: char,
    /// The character to use for the bottom-left corner of a multi-line label.
    /// Defaults to: `'╰'`.
    pub multiline_bottom_left_char: char,
    /// The character to use when marking the bottom of a multi-line label.
    /// Defaults to: `'─'`.
    pub multiline_bottom_char: char,
    /// The character to use for the left of a multi-line label.
    /// Defaults to: `'│'`.
    pub multiline_left_char: char,
    /// The character to use for the note bullet.
    /// Defaults to: `'='`.
    pub note_bullet_char: char,
}

impl Default for Config {
    fn default() -> Config {
        Config {
            display_style: DisplayStyle::Rich,
            tab_width: 4,
            styles: Styles::default(),
            source_border_top_left_char: '┌',
            source_border_top_char: '─',
            source_border_left_char: '│',
            primary_caret_char: '^',
            secondary_caret_char: '-',
            multiline_primary_caret_char: '^',
            multiline_secondary_caret_char: '\'',
            multiline_top_left_char: '╭',
            multiline_top_char: '─',
            multiline_bottom_left_char: '╰',
            multiline_bottom_char: '─',
            multiline_left_char: '│',
            note_bullet_char: '=',
        }
    }
}

impl Config {
    /// Measure the width of a string, taking into account the tab width.
    pub fn width(&self, s: &str) -> usize {
        use unicode_width::UnicodeWidthChar;

        s.chars()
            .map(|ch| match ch {
                '\t' => self.tab_width,
                _ => ch.width().unwrap_or(0),
            })
            .sum()
    }

    /// Get the amount of spaces we should use for printing tabs.
    pub fn tab_padding(&self) -> String {
        (0..self.tab_width).map(|_| ' ').collect()
    }
}

/// The display style to use when rendering diagnostics.
#[derive(Clone, Debug)]
pub enum DisplayStyle {
    /// Output a richly formatted diagnostic, with source code previews.
    ///
    /// ```text
    /// error[E0001]: unexpected type in `+` application
    ///
    ///    ┌── test:2:9 ───
    ///    │
    ///  2 │ (+ test "")
    ///    │         ^^ expected `Int` but found `String`
    ///    │
    ///    = expected type `Int`
    ///         found type `String`
    /// ```
    Rich,
    /// Output a short diagnostic, with a line number, severity, and message.
    ///
    /// ```text
    /// test:2:9: error[E0001]: unexpected type in `+` application
    /// ```
    Short,
}

/// Styles to use when rendering the diagnostic.
#[derive(Clone, Debug)]
pub struct Styles {
    /// The style to use when rendering bug headers.
    /// Defaults to `fg:red bold intense`.
    pub header_bug: ColorSpec,
    /// The style to use when rendering error headers.
    /// Defaults to `fg:red bold intense`.
    pub header_error: ColorSpec,
    /// The style to use when rendering warning headers.
    /// Defaults to `fg:yellow bold intense`.
    pub header_warning: ColorSpec,
    /// The style to use when rendering note headers.
    /// Defaults to `fg:green bold intense`.
    pub header_note: ColorSpec,
    /// The style to use when rendering help headers.
    /// Defaults to `fg:cyan bold intense`.
    pub header_help: ColorSpec,
    /// The style to use when the main diagnostic message.
    /// Defaults to `bold intense`.
    pub header_message: ColorSpec,

    /// The style to use when rendering bug labels.
    /// Defaults to `fg:red`.
    pub primary_label_bug: ColorSpec,
    /// The style to use when rendering error labels.
    /// Defaults to `fg:red`.
    pub primary_label_error: ColorSpec,
    /// The style to use when rendering warning labels.
    /// Defaults to `fg:yellow`.
    pub primary_label_warning: ColorSpec,
    /// The style to use when rendering note labels.
    /// Defaults to `fg:green`.
    pub primary_label_note: ColorSpec,
    /// The style to use when rendering help labels.
    /// Defaults to `fg:cyan`.
    pub primary_label_help: ColorSpec,
    /// The style to use when rendering secondary labels.
    /// Defaults `fg:blue` (or `fg:cyan` on windows).
    pub secondary_label: ColorSpec,

    /// The style to use when rendering the line numbers.
    /// Defaults `fg:blue` (or `fg:cyan` on windows).
    pub line_number: ColorSpec,
    /// The style to use when rendering the source code borders.
    /// Defaults `fg:blue` (or `fg:cyan` on windows).
    pub source_border: ColorSpec,
    /// The style to use when rendering the note bullets.
    /// Defaults `fg:blue` (or `fg:cyan` on windows).
    pub note_bullet: ColorSpec,
}

impl Styles {
    /// The style used to mark a header at a given severity.
    pub fn header(&self, severity: Severity) -> &ColorSpec {
        match severity {
            Severity::Bug => &self.header_bug,
            Severity::Error => &self.header_error,
            Severity::Warning => &self.header_warning,
            Severity::Note => &self.header_note,
            Severity::Help => &self.header_help,
        }
    }

    /// The style used to mark a primary label at a given severity.
    pub fn primary_label(&self, severity: Severity) -> &ColorSpec {
        match severity {
            Severity::Bug => &self.primary_label_bug,
            Severity::Error => &self.primary_label_error,
            Severity::Warning => &self.primary_label_warning,
            Severity::Note => &self.primary_label_note,
            Severity::Help => &self.primary_label_help,
        }
    }
}

impl Default for Styles {
    fn default() -> Styles {
        // Blue is really difficult to see on the standard windows command line
        #[cfg(windows)]
        const BLUE: Color = Color::Cyan;
        #[cfg(not(windows))]
        const BLUE: Color = Color::Blue;

        let header = ColorSpec::new().set_bold(true).set_intense(true).clone();

        Styles {
            header_bug: header.clone().set_fg(Some(Color::Red)).clone(),
            header_error: header.clone().set_fg(Some(Color::Red)).clone(),
            header_warning: header.clone().set_fg(Some(Color::Yellow)).clone(),
            header_note: header.clone().set_fg(Some(Color::Green)).clone(),
            header_help: header.clone().set_fg(Some(Color::Cyan)).clone(),
            header_message: header.clone(),

            primary_label_bug: ColorSpec::new().set_fg(Some(Color::Red)).clone(),
            primary_label_error: ColorSpec::new().set_fg(Some(Color::Red)).clone(),
            primary_label_warning: ColorSpec::new().set_fg(Some(Color::Yellow)).clone(),
            primary_label_note: ColorSpec::new().set_fg(Some(Color::Green)).clone(),
            primary_label_help: ColorSpec::new().set_fg(Some(Color::Cyan)).clone(),
            secondary_label: ColorSpec::new().set_fg(Some(BLUE)).clone(),

            line_number: ColorSpec::new().set_fg(Some(BLUE)).clone(),
            source_border: ColorSpec::new().set_fg(Some(BLUE)).clone(),
            note_bullet: ColorSpec::new().set_fg(Some(BLUE)).clone(),
        }
    }
}

/// A command line argument that configures the coloring of the output.
///
/// This can be used with command line argument parsers like `clap` or `structopt`.
///
/// # Example
///
/// ```rust
/// use structopt::StructOpt;
/// use codespan_reporting::term::termcolor::StandardStream;
/// use codespan_reporting::term::ColorArg;
///
/// #[derive(Debug, StructOpt)]
/// #[structopt(name = "groovey-app")]
/// pub struct Opts {
///     /// Configure coloring of output
///     #[structopt(
///         long = "color",
///         parse(try_from_str),
///         default_value = "auto",
///         raw(possible_values = "ColorArg::VARIANTS", case_insensitive = "true")
///     )]
///     pub color: ColorArg,
/// }
///
/// fn main() {
///     let opts = Opts::from_args();
///     let writer = StandardStream::stderr(opts.color.into());
/// }
/// ```
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct ColorArg(pub ColorChoice);

impl ColorArg {
    /// Allowed values the argument.
    ///
    /// This is useful for generating documentation via `clap` or `structopt`'s
    /// `possible_values` configuration.
    pub const VARIANTS: &'static [&'static str] = &["auto", "always", "ansi", "never"];
}

impl FromStr for ColorArg {
    type Err = &'static str;

    fn from_str(src: &str) -> Result<ColorArg, &'static str> {
        match src {
            _ if src.eq_ignore_ascii_case("auto") => Ok(ColorArg(ColorChoice::Auto)),
            _ if src.eq_ignore_ascii_case("always") => Ok(ColorArg(ColorChoice::Always)),
            _ if src.eq_ignore_ascii_case("ansi") => Ok(ColorArg(ColorChoice::AlwaysAnsi)),
            _ if src.eq_ignore_ascii_case("never") => Ok(ColorArg(ColorChoice::Never)),
            _ => Err("valid values: auto, always, ansi, never"),
        }
    }
}

impl Into<ColorChoice> for ColorArg {
    fn into(self) -> ColorChoice {
        self.0
    }
}
