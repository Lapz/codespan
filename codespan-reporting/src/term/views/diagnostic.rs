use std::io;
use termcolor::WriteColor;

use crate::diagnostic::Diagnostic;
use crate::files::Files;
use crate::term::Config;

use super::{Header, Locus, NewLine, SourceSnippet};

/// Output a richly formatted diagnostic, with source code previews.
pub struct RichDiagnostic<'a, FileId> {
    diagnostic: &'a Diagnostic<FileId>,
}

impl<'a, FileId> RichDiagnostic<'a, FileId>
where
    FileId: Copy + Ord,
{
    pub fn new(diagnostic: &'a Diagnostic<FileId>) -> RichDiagnostic<'a, FileId> {
        RichDiagnostic { diagnostic }
    }

    pub fn emit<'files>(
        &self,
        files: &'files impl Files<'files, FileId = FileId>,
        writer: &mut (impl WriteColor + ?Sized),
        config: &Config,
    ) -> io::Result<()>
    where
        FileId: 'files,
    {
        use std::collections::BTreeMap;

        use super::MarkStyle;

        Header::new(self.diagnostic).emit(writer, config)?;
        NewLine::new().emit(writer, config)?;

        let primary_label = &self.diagnostic.primary_label;
        let primary_file_id = self.diagnostic.primary_label.file_id;
        let severity = self.diagnostic.severity;
        let notes = &self.diagnostic.notes;

        // Group labels by file

        let mut label_groups = BTreeMap::new();

        label_groups
            .entry(primary_file_id)
            .or_insert(vec![])
            .push((primary_label, MarkStyle::Primary(severity)));

        for secondary_label in &self.diagnostic.secondary_labels {
            label_groups
                .entry(secondary_label.file_id)
                .or_insert(vec![])
                .push((secondary_label, MarkStyle::Secondary));
        }

        // Emit the snippets, starting with the one that contains the primary label

        let labels = label_groups.remove(&primary_file_id).unwrap_or(vec![]);
        SourceSnippet::new(primary_file_id, labels, notes).emit(files, writer, config)?;
        NewLine::new().emit(writer, config)?;

        for (file_id, labels) in label_groups {
            SourceSnippet::new(file_id, labels, &[]).emit(files, writer, config)?;
            NewLine::new().emit(writer, config)?;
        }

        Ok(())
    }
}

/// Output a short diagnostic, with a line number, severity, and message.
pub struct ShortDiagnostic<'a, FileId> {
    diagnostic: &'a Diagnostic<FileId>,
}

impl<'a, FileId> ShortDiagnostic<'a, FileId>
where
    FileId: Copy + Ord,
{
    pub fn new(diagnostic: &'a Diagnostic<FileId>) -> ShortDiagnostic<'a, FileId> {
        ShortDiagnostic { diagnostic }
    }

    pub fn emit<'files>(
        &self,
        files: &'files impl Files<'files, FileId = FileId>,
        writer: &mut (impl WriteColor + ?Sized),
        config: &Config,
    ) -> io::Result<()>
    where
        FileId: 'files,
    {
        let label = &self.diagnostic.primary_label;
        let start = label.range.start;

        let origin = files.origin(label.file_id).expect("origin");
        let line_index = files.line_index(label.file_id, start).expect("line_index");
        let line = files.line(label.file_id, line_index).expect("line");

        Locus::new(origin, line.number, line.column_number(start)).emit(writer, config)?;
        write!(writer, ": ")?;
        Header::new(self.diagnostic).emit(writer, config)?;

        Ok(())
    }
}
