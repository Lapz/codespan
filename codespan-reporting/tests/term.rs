use codespan::Files;
use codespan_reporting::diagnostic::{Diagnostic, Label, Title};
use codespan_reporting::term::{termcolor::Color, Config, Styles};

mod support;

use self::support::TestData;

fn test_config() -> Config {
    Config {
        // Always use blue so tests are consistent across platforms
        styles: Styles::with_blue(Color::Blue),
        ..Config::default()
    }
}

mod empty_spans {
    use super::*;

    lazy_static::lazy_static! {
        static ref TEST_DATA: TestData = {
            let mut files = Files::new();

            let file_id = files.add("hello", "Hello world!\nBye world!".to_owned());
            let eof = files.source_span(file_id).end();

            let diagnostics = vec![
                Diagnostic::note()
                    .with_title("middle")
                    .with_labels(vec![Label::primary(file_id, 6..6, "middle")]),
                Diagnostic::note()
                    .with_title("end of line")
                    .with_labels(vec![Label::primary(file_id, 12..12, "end of line")]),
                Diagnostic::note()
                    .with_title("end of file")
                    .with_labels(vec![Label::primary(file_id, eof..eof, "end of file")]),
            ];

            TestData { files, diagnostics }
        };
    }

    #[test]
    fn color() {
        insta::assert_snapshot!("color", TEST_DATA.emit_color(&test_config()));
    }

    #[test]
    fn no_color() {
        insta::assert_snapshot!("no_color", TEST_DATA.emit_no_color(&test_config()));
    }
}

mod multifile {
    use super::*;

    lazy_static::lazy_static! {
        static ref TEST_DATA: TestData = {
            let mut files = Files::new();

            let file_id1 = files.add(
                "Data/Nat.fun",
                unindent::unindent(
                    "
                        module Data.Nat where

                        data Nat : Type where
                            zero : Nat
                            succ : Nat → Nat

                        {-# BUILTIN NATRAL Nat #-}

                        infixl 6 _+_ _-_

                        _+_ : Nat → Nat → Nat
                        zero    + n₂ = n₂
                        succ n₁ + n₂ = succ (n₁ + n₂)

                        _-_ : Nat → Nat → Nat
                        n₁      - zero    = n₁
                        zero    - succ n₂ = zero
                        succ n₁ - succ n₂ = n₁ - n₂
                    ",
                ),
            );

            let file_id2 = files.add(
                "Test.fun",
                unindent::unindent(
                    r#"
                        module Test where

                        _ : Nat
                        _ = 123 + "hello"
                    "#,
                ),
            );

            let diagnostics = vec![
                // Unknown builtin error
                Diagnostic::error()
                    .with_title("unknown builtin: `NATRAL`")
                    .with_labels(vec![Label::primary(file_id1, 96..102, "unknown builtin")])
                    .with_notes(vec![
                        "there is a builtin with a similar name: `NATURAL`".to_owned(),
                    ]),
                // Unused parameter warning
                Diagnostic::warning()
                    .with_title("unused parameter pattern: `n₂`")
                    .with_labels(vec![Label::primary(file_id1, 285..289, "unused parameter")])
                    .with_notes(vec!["consider using a wildcard pattern: `_`".to_owned()]),
                // Unexpected type error
                Diagnostic::error()
                    .with_title(Title::new("unexpected type in application of `_+_`").with_code("E0001"))
                    .with_labels(vec![
                        Label::primary(file_id2, 37..44, "expected `Nat`, found `String`"),
                        Label::secondary(file_id1, 130..155, "based on the definition of `_+_`"),
                    ]),
            ];

            TestData { files, diagnostics }
        };
    }

    #[test]
    fn color() {
        insta::assert_snapshot!("color", TEST_DATA.emit_color(&test_config()));
    }

    #[test]
    fn no_color() {
        insta::assert_snapshot!("no_color", TEST_DATA.emit_no_color(&test_config()));
    }
}

mod fizz_buzz {
    use super::*;

    lazy_static::lazy_static! {
        static ref TEST_DATA: TestData = {
            let mut files = Files::new();

            let file_id = files.add(
                "FizzBuzz.fun",
                unindent::unindent(
                    r#"
                        module FizzBuzz where

                        fizz₁ : Nat → String
                        fizz₁ num = case (mod num 5) (mod num 3) of
                            0 0 => "FizzBuzz"
                            0 _ => "Fizz"
                            _ 0 => "Buzz"
                            _ _ => num

                        fizz₂ num =
                            case (mod num 5) (mod num 3) of
                                0 0 => "FizzBuzz"
                                0 _ => "Fizz"
                                _ 0 => "Buzz"
                                _ _ => num
                    "#,
                ),
            );

            let diagnostics = vec![
                // Incompatible match clause error
                Diagnostic::error()
                    .with_title(Title::new("`case` clauses have incompatible types").with_code("E0308"))
                    .with_labels(vec![
                        Label::primary(file_id, 163..166, "expected `String`, found `Nat`"),
                        Label::secondary(file_id, 62..166, "`case` clauses have incompatible types"),
                        Label::secondary(file_id, 41..47, "expected type `String` found here"),
                    ])
                    .with_notes(vec![unindent::unindent(
                        "
                        expected type `String`
                           found type `Nat`
                        ",
                    )]),
                // Incompatible match clause error
                Diagnostic::error()
                    .with_title(Title::new("`case` clauses have incompatible types").with_code("E0308"))
                    .with_labels(vec![
                        Label::primary(file_id, 303..306, "expected `String`, found `Nat`"),
                        Label::secondary(file_id, 186..306, "`case` clauses have incompatible types"),
                        Label::secondary(file_id, 233..243, "this is found to be of type `String`"),
                        Label::secondary(file_id, 259..265, "this is found to be of type `String`"),
                        Label::secondary(file_id, 281..287, "this is found to be of type `String`"),
                    ])
                    .with_notes(vec![unindent::unindent(
                        "
                        expected type `String`
                           found type `Nat`
                        ",
                    )]),
            ];

            TestData { files, diagnostics }
        };
    }

    #[test]
    fn color() {
        insta::assert_snapshot!("color", TEST_DATA.emit_color(&test_config()));
    }

    #[test]
    fn no_color() {
        insta::assert_snapshot!("no_color", TEST_DATA.emit_no_color(&test_config()));
    }
}

mod tabbed {
    use super::*;

    lazy_static::lazy_static! {
        static ref TEST_DATA: TestData = {
            let mut files = Files::new();

            let file_id = files.add(
                "tabbed",
                [
                    "Entity:",
                    "\tArmament:",
                    "\t\tWeapon: DogJaw",
                    "\t\tReloadingCondition:\tattack-cooldown",
                    "\tFoo: Bar",
                ]
                .join("\n"),
            );

            let diagnostics = vec![
                Diagnostic::warning()
                    .with_title("unknown weapon `DogJaw`")
                    .with_labels(vec![Label::primary(file_id, 29..35, "the weapon")]),
                Diagnostic::warning()
                    .with_title("unknown condition `attack-cooldown`")
                    .with_labels(vec![Label::primary(file_id, 58..73, "the condition")]),
                Diagnostic::warning()
                    .with_title("unknown field `Foo`")
                    .with_labels(vec![Label::primary(file_id, 75..78, "the field")]),
            ];

            TestData { files, diagnostics }
        };
    }

    #[test]
    fn tab_width_default_no_color() {
        let config = test_config();

        insta::assert_snapshot!(
            "tab_width_default_no_color",
            TEST_DATA.emit_no_color(&config)
        );
    }

    #[test]
    fn tab_width_3_no_color() {
        let config = Config {
            tab_width: 3,
            ..test_config()
        };

        insta::assert_snapshot!("tab_width_3_no_color", TEST_DATA.emit_no_color(&config));
    }

    #[test]
    fn tab_width_6_no_color() {
        let config = Config {
            tab_width: 6,
            ..test_config()
        };

        insta::assert_snapshot!("tab_width_6_no_color", TEST_DATA.emit_no_color(&config));
    }
}
