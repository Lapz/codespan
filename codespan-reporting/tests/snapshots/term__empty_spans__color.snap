---
source: codespan-reporting/tests/term.rs
expression: TEST_DATA.emit_color(&test_config())
---
{fg:Green bold bright}note{bold bright}: middle{/}

   {fg:Blue}┌{/}{fg:Blue}──{/} hello:1:7 {fg:Blue}───{/}
   {fg:Blue}│{/}
 {fg:Blue}1{/} {fg:Blue}│{/} Hello {fg:Green}{/}world!
   {fg:Blue}│{/}       {fg:Green}^ middle{/}
   {fg:Blue}│{/}

{fg:Green bold bright}note{bold bright}: end of line{/}

   {fg:Blue}┌{/}{fg:Blue}──{/} hello:1:13 {fg:Blue}───{/}
   {fg:Blue}│{/}
 {fg:Blue}1{/} {fg:Blue}│{/} Hello world!{fg:Green}{/}
   {fg:Blue}│{/}             {fg:Green}^ end of line{/}
   {fg:Blue}│{/}

{fg:Green bold bright}note{bold bright}: end of file{/}

   {fg:Blue}┌{/}{fg:Blue}──{/} hello:2:11 {fg:Blue}───{/}
   {fg:Blue}│{/}
 {fg:Blue}2{/} {fg:Blue}│{/} Bye world!{fg:Green}{/}
   {fg:Blue}│{/}           {fg:Green}^ end of file{/}
   {fg:Blue}│{/}


